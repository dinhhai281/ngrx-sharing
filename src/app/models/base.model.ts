export interface RepsonseData<T> {
  result: null | T;
  message?: string;
}
