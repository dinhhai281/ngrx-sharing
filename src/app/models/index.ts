export * from './post.model';
export * from './user.model';
export * from './base.model';
export * from './collection.model';
