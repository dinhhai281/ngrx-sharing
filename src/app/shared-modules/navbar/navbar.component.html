<ng-container *ngIf="state.vm$ | async as vm">
  <div class="flex py-3 px-8 justify-between items-center bg-gray-800 shadow-sm fixed w-full top-0 z-10">
    <section class="flex items-center">
      <span
        class="text-pink-600 flex items-center justify-center text-xl font-semibold"
      >
        <ng-content select="[app-label]"></ng-content>
      </span>
    </section>
    <section class="flex justify-center items-center">
      <button class="mr-4 focus:outline-none flex items-center" (click)="onAddCollection()">
        <mat-icon svgIcon="add"></mat-icon>
      </button>
      <button class="focus:outline-none flex items-center"
        (click)="onToggleUserMenu()"
        cdkOverlayOrigin
        #userMenu="cdkOverlayOrigin"
      >
        <span
          class="h-8 w-8 rounded-full text-pink-700 bg-white border-2 border-gray-400 flex items-center justify-center text-xs"
          data-testid="avatar"
        >
          {{ vm.me | userAvatar:'NA' | uppercase }}
        </span>
      </button>
    </section>
  </div>
  <ng-template
    cdkConnectedOverlay
    [cdkConnectedOverlayOrigin]="userMenu"
    [cdkConnectedOverlayOpen]="vm.isOpenUserMenu"
    [cdkConnectedOverlayOffsetY]="8"
    (overlayOutsideClick)="onOverlayOutsideClick($event)"
  >
    <section class="flex flex-col rounded-lg bg-gray-800 w-48 border border-gray-900">
      <div class="p-3 text-gray-200 flex flex-col border-b border-gray-900">
        <span>Signed in as</span>
        <span class="font-semibold text-gray-100">{{ vm.me.email }}</span>
      </div>

      <ul class="py-3 text-gray-200 flex flex-col">
        <li class="px-3 py-1 cursor-pointer hover:bg-blue-400" (click)="onLogout()">Logout</li>
      </ul>
    </section>
  </ng-template>
</ng-container>


