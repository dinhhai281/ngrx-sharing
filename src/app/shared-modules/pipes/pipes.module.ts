import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { UserAvatarPipe } from './user-avatar.pipe';

@NgModule({
  declarations: [UserAvatarPipe],
  imports: [CommonModule],
  exports: [UserAvatarPipe],
})
export class PipesModule {}
