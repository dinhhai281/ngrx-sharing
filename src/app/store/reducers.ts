import * as fromPost from '@store/post';
import * as fromAuth from '@store/auth';
import * as fromCollection from '@store/collection';

export interface State {
  [fromPost.POST_FEATURE_KEY]?: fromPost.PostEntityState;
  [fromAuth.AUTH_FEATURE_KEY]?: fromAuth.AuthState;
  [fromCollection.COLLECTION_FEATURE_KEY]?: fromCollection.CollectionState;
}

export { fromPost, fromAuth, fromCollection };
