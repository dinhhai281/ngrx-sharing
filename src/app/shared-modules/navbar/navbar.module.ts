import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MatBadgeModule } from '@angular/material/badge';
import { MatDialogModule } from '@angular/material/dialog';
import { MatIconModule } from '@angular/material/icon';

import { PipesModule } from '../pipes/pipes.module';
import { NavbarComponent } from './navbar.component';

@NgModule({
  declarations: [NavbarComponent],
  imports: [
    CommonModule,
    MatBadgeModule,
    MatIconModule,
    OverlayModule,
    PipesModule,
    MatDialogModule,
  ],
  exports: [NavbarComponent],
})
export class NavbarModule {}
