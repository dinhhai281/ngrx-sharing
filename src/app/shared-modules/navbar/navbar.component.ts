import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';

import { NavbarStore } from './navbar.store';

@Component({
  selector: 'app-navbar',
  templateUrl: './navbar.component.html',
  styleUrls: ['./navbar.component.scss'],
  encapsulation: ViewEncapsulation.None,
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [NavbarStore],
})
export class NavbarComponent {
  constructor(public readonly state: NavbarStore) {}

  onLogout() {
    this.state.logout();
  }

  onAddCollection() {
    this.state.addCollection();
  }

  onToggleUserMenu() {
    this.state.toggleUserMenu();
  }

  onOverlayOutsideClick(event: Event) {
    event.stopImmediatePropagation();
    this.state.toggleUserMenu();
  }
}
