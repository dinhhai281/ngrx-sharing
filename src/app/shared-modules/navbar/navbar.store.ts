import { Injectable } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import { CollectionDialogComponent } from '@components/collection-dialog';
import { User } from '@app/models';
import { ComponentStore } from '@ngrx/component-store';
import { select, Store } from '@ngrx/store';
import { fromAuth, fromCollection, State } from '@store/reducers';
import { map, switchMap, tap, withLatestFrom } from 'rxjs/operators';

export interface NavbarState {
  me: User;
  isOpenUserMenu: boolean;
}

@Injectable()
export class NavbarStore extends ComponentStore<NavbarState> {
  constructor(private store: Store<State>, private matDialog: MatDialog) {
    super({ me: null, isOpenUserMenu: false });
  }

  readonly me$ = this.store.pipe(select(fromAuth.selectMe));
  readonly isOpenUserMenu$ = this.select(state => state.isOpenUserMenu);
  readonly vm$ = this.select(this.me$, this.isOpenUserMenu$, (me, isOpenUserMenu) => ({
    me,
    isOpenUserMenu,
  }));

  readonly setMe = this.updater<User>((state, me) => ({
    ...state,
    me,
  }));
  readonly toggleUserMenu = this.updater<void>(state => ({
    ...state,
    isOpenUserMenu: !state.isOpenUserMenu,
  }));

  readonly logout = this.effect(logoutEvent$ =>
    logoutEvent$.pipe(map(fromAuth.logout), tap(this.store.dispatch.bind(this.store))),
  );

  readonly addCollection = this.effect(addCollectionEvent$ =>
    addCollectionEvent$.pipe(
      withLatestFrom(this.me$),
      switchMap(([, me]) => {
        const dialog = this.matDialog.open(CollectionDialogComponent);
        return dialog.componentInstance.value$.pipe(
          tap(() => dialog.close()),
          map(collectionName =>
            fromCollection.createCollection({ collectionName, userId: me.token }),
          ),
        );
      }),
      tap(this.store.dispatch.bind(this.store)),
    ),
  );
}
