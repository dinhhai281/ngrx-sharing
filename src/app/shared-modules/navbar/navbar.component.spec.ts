import { OverlayModule } from '@angular/cdk/overlay';
import { TestBed } from '@angular/core/testing';
import { MatDialogModule } from '@angular/material/dialog';
import { MatIconModule, MatIconRegistry } from '@angular/material/icon';
import { FakeMatIconRegistry } from '@angular/material/icon/testing';
import { User } from '@app/models';
import { MockStore, provideMockStore } from '@ngrx/store/testing';
import { createInitialState } from '@store/auth';
import { State } from '@store/reducers';
import { render, RenderResult } from '@testing-library/angular';

import { PipesModule } from '../pipes/pipes.module';
import { UserAvatarPipe } from '../pipes/user-avatar.pipe';
import { NavbarComponent } from './navbar.component';
import { NavbarStore } from './navbar.store';

describe('NavbarComponent', () => {
  let component: RenderResult<NavbarComponent>;
  const testUser: User = { token: 'test', email: 'Test User' };
  const initialState: State = {
    auth: { ...createInitialState(), me: { token: 'test', email: 'Test User' } },
  };
  let store: MockStore<State>;
  let componentStore: NavbarStore;

  beforeEach(async () => {
    component = await render(NavbarComponent, {
      imports: [MatIconModule, OverlayModule, PipesModule, MatDialogModule],
      providers: [
        { provide: MatIconRegistry, useClass: FakeMatIconRegistry },
        NavbarStore,
        provideMockStore({ initialState }),
      ],
    });
    store = TestBed.inject(MockStore);
    componentStore = TestBed.inject(NavbarStore);
  });

  it('should match snapshot', () => {
    expect(component.container).toMatchSnapshot();
  });

  it('should render avatar with username', () => {
    const userAvatarPipe = new UserAvatarPipe();
    expect(component.getByTestId('avatar').textContent.trim()).toEqual(
      userAvatarPipe.transform(testUser).toUpperCase(),
    );
  });
});
