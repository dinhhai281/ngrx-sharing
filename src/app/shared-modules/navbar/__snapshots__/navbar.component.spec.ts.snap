// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavbarComponent should match snapshot 1`] = `
<div
  id="root0"
>
  
  
  <div
    class="flex py-3 px-8 justify-between items-center bg-gray-800 shadow-sm fixed w-full top-0 z-10"
  >
    <section
      class="flex items-center"
    >
      <span
        class="text-pink-600 flex items-center justify-center text-xl font-semibold"
      />
    </section>
    <section
      class="flex justify-center items-center"
    >
      <button
        class="mr-4 focus:outline-none flex items-center"
      >
        <mat-icon
          aria-hidden="true"
          class="mat-icon notranslate mat-icon-no-color"
          data-mat-icon-name="add"
          data-mat-icon-type="svg"
          role="img"
          svgicon="add"
        >
          <svg
            class="fake-testing-svg"
            fit=""
            focusable="false"
            height="100%"
            preserveAspectRatio="xMidYMid meet"
            width="100%"
          />
        </mat-icon>
      </button>
      <button
        cdkoverlayorigin=""
        class="focus:outline-none flex items-center"
      >
        <span
          class="h-8 w-8 rounded-full text-pink-700 bg-white border-2 border-gray-400 flex items-center justify-center text-xs"
          data-testid="avatar"
        >
           TE 
        </span>
      </button>
    </section>
  </div>
  
  
</div>
`;
